### PR TITLE
[SOL] New encoding for memory instructions

### PR DESCRIPTION
**Pull request by LucasSte**
[SIMD-0173](https://github.com/solana-foundation/solana-improvement-documents/pull/173) introduces a new encoding for memory instructions, as follows:

```
STB  => 0x27 
STH  => 0x37
STW  => 0x87
STDW => 0x97
```

```
SDXB  => 0x2f
SDXH  => 0x3f
SDXW  => 0x8f
SDXDW => 0x9f
```

```
LDXB  => 0x2c 
LDXH  => 0x3c
LDXW  => 0x8c
LDXDW => 0x9c
```

This PR brings this new encoding to the SBF target.

### CODE REVIEW

File: llvm/test/MC/Disassembler/SBF/sbf-ldst-new.txt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+0x3f,0x82,0x20,0x00,0x00,0x00,0x00,0x00
+0x8f,0x43,0x40,0x00,0x00,0x00,0x00,0x00
+0x9f,0x04,0x80,0x00,0x00,0x00,0x00,0x00
+
+
+
+# CHECK32-NEW: stxb [r1 + 17], w6
+# CHECK32-NEW: stxb [r10 - 3], w1
+# CHECK32-NEW: stxh [r2 + 32], w8
+# CHECK32-NEW: stxw [r3 + 64], w4
```
Comment (Lichtso) [inline llvm/test/MC/Disassembler/SBF/sbf-ldst-new.txt:47]: Missing `stxdw`

File: llvm/test/MC/Disassembler/SBF/sbf-ldst-new.txt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+0x3c,0x28,0x20,0x00,0x00,0x00,0x00,0x00
+0x8c,0x34,0x40,0x00,0x00,0x00,0x00,0x00
+0x9c,0x40,0x80,0x00,0x00,0x00,0x00,0x00
+
+
+
+# CHECK32-NEW: ldxb w6, [r1 + 17]
+# CHECK32-NEW: ldxb w1, [r10 - 3]
+# CHECK32-NEW: ldxh w8, [r2 + 32]
+# CHECK32-NEW: ldxw w4, [r3 + 64]
```
Comment (Lichtso) [inline llvm/test/MC/Disassembler/SBF/sbf-ldst-new.txt:23]: Missing `ldxdw`

File: llvm/test/MC/SBF/sbf-ldst-new.s [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+
+# CHECK32-OBJ-NEW: ldxb w6, [r1 + 0x11]
+# CHECK32-OBJ-NEW: ldxb w1, [r10 - 0x3]
+# CHECK32-OBJ-NEW: ldxh w8, [r2 + 0x20]
+# CHECK32-OBJ-NEW: ldxw w4, [r3 + 0x40]
+# CHECK32-ASM-NEW: encoding: [0x2c,0x16,0x11,0x00,0x00,0x00,0x00,0x00]
+# CHECK32-ASM-NEW: encoding: [0x2c,0xa1,0xfd,0xff,0x00,0x00,0x00,0x00]
+# CHECK32-ASM-NEW: encoding: [0x3c,0x28,0x20,0x00,0x00,0x00,0x00,0x00]
+# CHECK32-ASM-NEW: encoding: [0x8c,0x34,0x40,0x00,0x00,0x00,0x00,0x00]
```
Comment (Lichtso) [inline llvm/test/MC/SBF/sbf-ldst-new.s:39]: missing `ldxdw`

File: llvm/test/MC/SBF/sbf-ldst-new.s [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+
+# CHECK32-OBJ-NEW: stxb [r1 + 0x11], w6
+# CHECK32-OBJ-NEW: stxb [r10 - 0x3], w1
+# CHECK32-OBJ-NEW: stxh [r2 + 0x20], w8
+# CHECK32-OBJ-NEW: stxw [r3 + 0x40], w4
+# CHECK32-ASM-NEW: encoding: [0x2f,0x61,0x11,0x00,0x00,0x00,0x00,0x00]
+# CHECK32-ASM-NEW: encoding: [0x2f,0x1a,0xfd,0xff,0x00,0x00,0x00,0x00]
+# CHECK32-ASM-NEW: encoding: [0x3f,0x82,0x20,0x00,0x00,0x00,0x00,0x00]
+# CHECK32-ASM-NEW: encoding: [0x8f,0x43,0x40,0x00,0x00,0x00,0x00,0x00]
```
Comment (Lichtso) [inline llvm/test/MC/SBF/sbf-ldst-new.s:72]: missing `stxdw`

File: llvm/test/MC/Disassembler/SBF/sbf-ldst-new.txt (hunk 1 above)
Comment (LucasSte) [reply llvm/test/MC/Disassembler/SBF/sbf-ldst-new.txt:47]: `CHECK32` tests loads for the 32-bit register class, so `stxdw` isn't available here.

Comment (LucasSte) [reply llvm/test/MC/Disassembler/SBF/sbf-ldst-new.txt:47]: Have a look at line 35.

File: llvm/test/MC/Disassembler/SBF/sbf-ldst-new.txt (hunk 2 above)
Comment (LucasSte) [reply llvm/test/MC/Disassembler/SBF/sbf-ldst-new.txt:23]: They are at line 11. Here, we test for the 32-bit register class, for which `ldxdw` isn't available.

File: llvm/test/MC/SBF/sbf-ldst-new.s (hunk 3 above)
Comment (LucasSte) [reply llvm/test/MC/SBF/sbf-ldst-new.s:39]: `CHECK32` (same drill as previous comment).

File: llvm/test/MC/SBF/sbf-ldst-new.s (hunk 4 above)
Comment (LucasSte) [reply llvm/test/MC/SBF/sbf-ldst-new.s:72]: Same drill again.

**Review by Lichtso — APPROVED** [review]

**Pull request merged**